cura/UI/MachineSettingsManager.py

```diff
@@ -0,0 +1,72 @@
+
+from typing import Any, Dict, Optional, TYPE_CHECKING
+
+from PyQt5.QtCore import Qt, QObject, pyqtSlot
+
+from UM.i18n import i18nCatalog
+
+
+class MachineSettingsManager(QObject):
+
+    def __init__(self, parent: Optional["QObject"] = None) -> None:
+        super().__init__(parent)
+        self._i18n_catalog = i18nCatalog("cura")
+
+        from cura.CuraApplication import CuraApplication
+        self._application = CuraApplication.getInstance()
+
+    # Force rebuilding the build volume by reloading the global container stack. This is a bit of a hack, but it seems
+    # quite enough.
+    @pyqtSlot()
+    def forceUpdate(self) -> None:
+        self._application.getMachineManager().globalContainerChanged.emit()
+
+    # Function for the Machine Settings panel (QML) to update the compatible material diameter after a user has changed
+    # an extruder's compatible material diameter. This ensures that after the modification, changes can be notified
+    # and updated right away.
+    @pyqtSlot(int)
+    def updateMaterialForDiameter(self, extruder_position: int) -> None:
+        # Updates the material container to a material that matches the material diameter set for the printer
+        self._application.getMachineManager().updateMaterialWithVariant(str(extruder_position))
+
+    @pyqtSlot(int)
+    def setMachineExtruderCount(self, extruder_count: int) -> None:
+        # Note: this method was in this class before, but since it's quite generic and other plugins also need it
+        # it was moved to the machine manager instead. Now this method just calls the machine manager.
+        self._application.getMachineManager().setActiveMachineExtruderCount(extruder_count)
+
+    # FIXME(Lipu): Better document what this function does, especially the fuzzy gcode flavor and has_materials logic
```

Comment:
There is a fixme, so that should be fixed before merging.

Reply:
I actually don't remember why it's done like this. That's why I put the `FIXME` here.

Reply:
The Ultimaker 2 family does not have materials in Cura by default, because the material is to be set on the printer. But when switching to Marlin flavor, the printer firmware can not change/insert material settings on the fly so they need to be configured in Cura. So when switching between gcode flavors, materials may need to be enabled/disabled.

Reply:
@fieldOfView Thanks for the explanation. I'll update the doc.

Reply:
Updated in 109631d02

Reply:
If I had known you would quote me word for word, I would have worded it more carefully, but I guess this works. And its cool to be quoted in the code ;-)